src/taskorchestrationexecutor.ts

```
@@ -0,0 +1,493 @@
+            [HistoryEventType.TaskFailed]: [false, "TaskScheduledId"],
+            [HistoryEventType.SubOrchestrationInstanceFailed]: [false, "TaskScheduledId"],
+        };
+        this.initialize();
+    }
+
+    /**
+     * @hidden
+     *
+     * Initialize the task orchestration executor for a brand new orchestrator invocation.
+     * To be called in ContinueAsNew scenarios as well.
+     */
+    private initialize(): void {
+        // The very first task, to kick-start the generator, is just a dummy/no-op task
+        this.currentTask = new InternalOnlyTask();
+        this.currentTask.SetValue(false, undefined);
+
+        this.sequenceNumber = 0;
+        this.willContinueAsNew = false;
+        this.openTasks = {};
+        this.openEvents = {};
+        this.actions = [];
+        this.deferredTasks = {};
+
+        this.output = undefined;
+        this.exception = undefined;
+        this.orchestratorReturned = false;
+    }
+
+    /**
+     * @hidden
+     *
+     * Start an orchestration's execution, replaying based on the currently-available History.
+     *
+     * @param context
+     *  The orchestration context
+     * @param history
+     *  The orchestration history
+     * @param schemaVersion
+     *  The OOProc output schema version expected by the DF extension
+     * @param fn
+     *  The user-code defining the orchestration
+     *
+     * @returns
+     *  Returns void but communicates the resulting orchestrator state via the context object's handler
+     */
+    public async execute(
+        context: IOrchestrationFunctionContext,
+        history: HistoryEvent[],
+        schemaVersion: UpperSchemaVersion,
+        fn: (context: IOrchestrationFunctionContext) => IterableIterator<unknown>
+    ): Promise<void> {
+        this.schemaVersion = schemaVersion;
+        this.context = context.df;
+        this.generator = fn(context) as Generator<TaskBase, any, any>; // TODO: what happens if code is not a generator?
```

Regarding this TODO, should we throw an error if it's not a generator?

---

We actually support non-generators in Python. For example, a trivial orchestration that just returns "Hello!" is a valid orchestration.

---

Oh interesting. I guess if we wanted to also support that in Python we could just set `fn` as the orchestrator output value if it's not a generator.

---

I tested this on a newly added unit test, and it seems the current implementation already supports non-generator orchestrations; which is exactly what we want!

I've removed this TODO comment